packages/relayer-lib/src/tx_builder/eth_to_cosmos.rs

```diff
@@ -399,39 +340,105 @@ where
+        };
+
+        let target_height = Height {
+            revision_number: 0,
+            revision_height: minimum_block_number,
+        };
+
+        let now_since_unix = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH)?;
+
+        let mut timeout_msgs = cosmos::target_events_to_timeout_msgs(
+            dest_events,
+            &target_client_id,
+            &target_height,
+            &self.signer_address,
+            now_since_unix.as_secs(),
+        );
+
+        let (mut recv_msgs, mut ack_msgs) = cosmos::src_events_to_recv_and_ack_msgs(
+            src_events,
+            &target_client_id,
+            &target_height,
+            &self.signer_address,
+            now_since_unix.as_secs(),
+        );
+
+        let mut ethereum_client_state =
+            self.ethereum_client_state(target_client_id.clone()).await?;
+        let mut ethereum_consensus_state = self
+            .ethereum_consensus_state(target_client_id.clone(), 0)
+            .await?;
+
+        tracing::info!(
+            "Relaying events from Ethereum to Cosmos for client {}, target block number: {}, client state latest slot: {}, consensus state slot: {}, current sync committee: {:?}, next sync committee: {:?}",
+            target_client_id,
+            minimum_block_number,
+            ethereum_consensus_state.slot,
+            ethereum_consensus_state.slot,
+            ethereum_consensus_state.current_sync_committee,
+            ethereum_consensus_state.next_sync_committee,
+        );
+
+        // get updates if necessary
+        let headers = if minimum_block_number > ethereum_consensus_state.execution_block_number {
+            self.wait_for_light_client_readiness(minimum_block_number)
+                .await?;
+            // Update the client state and consensus state, in case they have changed while we were waiting
+            ethereum_client_state = self.ethereum_client_state(target_client_id.clone()).await?;
+            ethereum_consensus_state = self
+                .ethereum_consensus_state(target_client_id.clone(), 0)
+                .await?;
+            self.get_update_headers(&ethereum_client_state, &ethereum_consensus_state)
+                .await?
+        } else {
+            vec![]
+        };
+
+        let proof_slot = headers.last().map_or(ethereum_consensus_state.slot, |h| {
+            h.consensus_update.finalized_header.beacon.slot
+        });
 
         cosmos::inject_ethereum_proofs(
             &mut recv_msgs,
             &mut ack_msgs,
             &mut timeout_msgs,
             &self.eth_client,
+            &self.beacon_api_client,
             &ethereum_client_state.ibc_contract_address.to_string(),
             ethereum_client_state.ibc_commitment_slot,
-            proof_block_number,
-            latest_trusted_slot,
+            proof_slot,
         )
         .await?;
 
         let update_msgs = headers
+            .clone()
             .into_iter()
```

Comment:
You don't have to use clone here
```suggestion
            .iter()
```